### PR TITLE
Add metric for Thrift connector remote requests

### PR DESCRIPTION
**Pull request by hellium01**
Add stats for retryDriver:
successRate
failureRate
total attempt
number of attempts per succeed request
time spent per succeed request
number of attempts per failed request
time spent per failed request

### CODE REVIEW

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 1]
```diff
@@ -125,15 +191,23 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
...
+                getStats(callableName, stats -> stats.updateSucceedLatency(retryStatus.getDuration()));
+                return result;
             }
             catch (InterruptedException ie) {
                 Thread.currentThread().interrupt();
                 throw propagate(ie);
             }
             catch (Exception e) {
                 if (retryStatus.shouldStopRetry(e)) {
+                    getStats(callableName, stats -> stats.updateFinalFailure());
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: can be replaced with method reference. Same in a few other places

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 2]
```diff
@@ -125,15 +191,23 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
 
         while (true) {
             retryStatus.nextAttempt();
+            getStats(callableName, stats -> stats.updateAttempt());
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Let's not add the number of total attempts for now as it's possible to get it as a sum of the other two.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 3]
```diff
@@ -125,15 +191,23 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
...
             try {
-                return callable.call();
+                V result = callable.call();
+                getStats(callableName, stats -> stats.updateSuccess());
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: I'd add two methods: `recordMetricsOnSuccess()` and `recordMetricsOnFailure()`. It will make the code more compact and descriptive and will be a bit more efficient, because you can lookup the stats once and won't need lambda.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java (hunk 1 above)
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Let's record failure stats in `propagate` in order not to miss some of them, for example those caused by interrupt.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 4]
```diff
@@ -58,7 +73,9 @@ private RetryDriver(
             Optional<Runnable> retryRunnable,
             Predicate<Exception> stopRetrying,
             Function<Exception, Exception> classifier,
-            ListeningScheduledExecutorService retryExecutorService)
+            ListeningScheduledExecutorService retryExecutorService,
+            LoadingCache<String, RetryStats> methodStats,
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: See if you can replace cache with a (concurrent) map.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 5]
```diff
@@ -49,13 +50,13 @@
     private final RetryDriver retry;
 
     @Inject
-    public RetryingPrestoThriftServiceProvider(@NonRetrying PrestoThriftServiceProvider original, @ForRetryDriver ListeningScheduledExecutorService retryExecutor, ThriftConnectorConfig config)
+    public RetryingPrestoThriftServiceProvider(@NonRetrying PrestoThriftServiceProvider original, @ForRetryDriver ListeningScheduledExecutorService retryExecutor, final MBeanExporter mBeanExporter, ThriftConnectorConfig config)
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: This is quite long (over 200 characters), please wrap arguments

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 6]
```diff
@@ -81,37 +100,84 @@ private RetryDriver(ListeningScheduledExecutorService executor)
...
+                            throws Exception
+                    {
+                        RetryStats methodStats = new RetryStats();
+                        String generatedMBeanName = generatedNameOf(RetryStats.class, methodName);
+                        try {
+                            mBeanExporter.unexport(generatedMBeanName);
+                        }
+                        catch (JmxException ignored) {
+                        }
+                        mBeanExporter.export(generatedMBeanName, methodStats);
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Exporting on every call is not the right approach. You want to have a single stats class covering every method which is exported as normal during the Guice binding. Take a look at `ThriftHiveMetastoreStats` and how it is used by `ThriftHiveMetastore`.

**Comment by electrum** [comment]
The commit message can be improved. How about "Add metrics for Thrift connector RetryDriver"

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 7]
```diff
@@ -48,6 +49,12 @@
     private static final Logger log = Logger.get(RetryingPrestoThriftServiceProvider.class);
     private final PrestoThriftServiceProvider original;
     private final RetryDriver retry;
+    private static final String LIST_SCHEMA_METHOD = "listSchemaNames";
```
Comment (jessesleeping) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: We can do something like passing a `RetryStats` object to the `RetryDriver.run()` function. And keep a `RetryStats` object per thrift method inside `RetryingPrestoThriftServiceProvider`. By doing this we can get rid of all these string constants and the concurrent hash map.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java (hunk 6 above)
Comment (hellium01) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Thanks for reviewing @electrum , I updated the commit to use a stand alone object for exporting the stats, please let me know if that works better.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 8]
```diff
@@ -80,13 +85,20 @@ private static Exception classifyException(Exception e)
...
 
     @Override
     public PrestoThriftService selectedHostClient(List<HostAddress> hosts)
     {
-        return new RetryingService(() -> original.selectedHostClient(hosts), retry);
+        return new RetryingService(() -> original.selectedHostClient(hosts), retry, retryingPrestoThriftServiceStats);
+    }
+
+    @Managed
+    @Nested
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: I think you want this flattened, so the stats are directly on this class.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 9]
```diff
@@ -58,6 +62,7 @@ public RetryingPrestoThriftServiceProvider(@NonRetrying PrestoThriftServiceProvi
         requireNonNull(retryExecutor, "retryExecutor is null");
         requireNonNull(config, "config is null");
 
+        retryingPrestoThriftServiceStats = new RetryingPrestoThriftServiceStats();
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: Initialize in the field declaration. Alternatively, you could inject this and export it directly, removing the need for the method below.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 10]
```diff
@@ -50,6 +53,7 @@
     private static final Logger log = Logger.get(RetryingPrestoThriftServiceProvider.class);
     private final PrestoThriftServiceProvider original;
     private final RetryDriver retry;
+    private final RetryingPrestoThriftServiceStats retryingPrestoThriftServiceStats;
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: This name is very long for no reason. It can be called “stats”

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 11]
```diff
@@ -177,4 +191,56 @@ public void close()
             client = null;
         }
     }
+
+    public static class RetryingPrestoThriftServiceStats
+    {
+        private final RetryStats listSchemaNameStats = new RetryStats();
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: Remove “stats” from all these

Comment (electrum) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: See ThriftHiveMetastoreStats

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 12]
```diff
@@ -117,7 +124,7 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
     /**
      * This is only for sync calls. For Callable that returns Future object, use runAsync instead.
      */
-    public <V> V run(String callableName, Callable<V> callable)
+    public <V> V run(String callableName, RetryStats retryStats, Callable<V> callable)
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Let’s call this “stats”

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 13]
```diff
@@ -126,15 +133,19 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
         while (true) {
             retryStatus.nextAttempt();
             try {
-                return callable.call();
+                V result = callable.call();
+                retryStats.updateSuccess();
+                retryStats.updateSucceedAttempts(retryStatus.attempts.get());
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Use a getter rather than direct field access

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 14]
```diff
@@ -143,29 +154,32 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
...
     }
 
-    private <V> ListenableFuture<V> runAsyncInternal(String callableName, Callable<ListenableFuture<V>> callable, RetryStatus retryStatus)
+    private <V> ListenableFuture<V> runAsyncInternal(String callableName, RetryStats retryStats, Callable<ListenableFuture<V>> callable, RetryStatus retryStatus)
     {
         retryStatus.nextAttempt();
         ListenableFuture<V> resultFuture;
         try {
             resultFuture = callable.call();
+            retryStats.updateSuccess();
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: I’d name all these “add” since “update” implies we are setting (overwriting) the value.

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java [hunk 15]
```diff
@@ -75,7 +76,8 @@ void testNoRetry()
             throws Exception
     {
         SyncTaskImpl task = new SyncTaskImpl(SUCCESS);
-        Integer result = retry.run("test", task);
+        RetryStats retryStats = new RetryStats();
+        Integer result = retry.run("test", retryStats, task);
```
Comment (electrum) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java]: Does this need to be boxed (can it be null)?

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 16]
```diff
@@ -209,6 +226,11 @@ long getRetryDelayInMs()
             return delayInMs + jitter;
         }
 
+        long getDuration()
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Where is this used? If needed, we should return the Duration, rather than returning a long with an unknown unit.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 17]
```diff
@@ -226,4 +248,79 @@ void nextAttempt()
             }
         }
     }
+
+    public static final class RetryStats
+    {
+        private final CounterStat success = new CounterStat();
+        private final CounterStat finalFailure = new CounterStat();
+        private final DistributionStat failureRetryCountDistribution = new DistributionStat();
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Remove “distribution” from the names

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java (hunk 15 above)
Comment (hellium01) [reply presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java]: Allowing null will require adding checks to skip updating the stats.  Maybe creating a unused object here is easier?

File: presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java [hunk 18]
```diff
@@ -1528,61 +1528,61 @@ public void setup()
 
         SchemaTableName table1 = new SchemaTableName("s1", "t1");
         inSetupTransaction(session -> metadata.createTable(session, TPCH_CATALOG, new ConnectorTableMetadata(table1,
-                ImmutableList.of(
-                        new ColumnMetadata("a", BIGINT),
-                        new ColumnMetadata("b", BIGINT),
-                        new ColumnMetadata("c", BIGINT),
-                        new ColumnMetadata("d", BIGINT))),
+                        ImmutableList.of(
```
Comment (AlekseiS) [inline presto-main/src/test/java/com/facebook/presto/sql/analyzer/TestAnalyzer.java]: Looks like changes in this file are unrelated, let's not do them.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 19]
```diff
@@ -50,14 +53,16 @@
...
 
     @Inject
-    public RetryingPrestoThriftServiceProvider(@NonRetrying PrestoThriftServiceProvider original, @ForRetryDriver ListeningScheduledExecutorService retryExecutor, ThriftConnectorConfig config)
+    public RetryingPrestoThriftServiceProvider(@NonRetrying PrestoThriftServiceProvider original, @ForRetryDriver ListeningScheduledExecutorService retryExecutor, RetryingPrestoThriftServiceStats stats, ThriftConnectorConfig config)
     {
         this.original = requireNonNull(original, "original is null");
         requireNonNull(retryExecutor, "retryExecutor is null");
         requireNonNull(config, "config is null");
 
+        this.stats = stats;
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: `requireNonNull`

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 20]
```diff
@@ -50,14 +53,16 @@
     private static final Logger log = Logger.get(RetryingPrestoThriftServiceProvider.class);
     private final PrestoThriftServiceProvider original;
     private final RetryDriver retry;
+    private final RetryingPrestoThriftServiceStats stats;
 
     @Inject
-    public RetryingPrestoThriftServiceProvider(@NonRetrying PrestoThriftServiceProvider original, @ForRetryDriver ListeningScheduledExecutorService retryExecutor, ThriftConnectorConfig config)
+    public RetryingPrestoThriftServiceProvider(@NonRetrying PrestoThriftServiceProvider original, @ForRetryDriver ListeningScheduledExecutorService retryExecutor, RetryingPrestoThriftServiceStats stats, ThriftConnectorConfig config)
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: Why not add the new parameter to the end of the list?

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 21]
```diff
@@ -96,11 +101,13 @@ public PrestoThriftService selectedHostClient(List<HostAddress> hosts)
         private final Supplier<PrestoThriftService> clientSupplier;
         private final RetryDriver retry;
         private PrestoThriftService client;
+        private RetryingPrestoThriftServiceStats retryingPrestoThriftServiceStats;
 
-        public RetryingService(Supplier<PrestoThriftService> clientSupplier, RetryDriver retry)
+        public RetryingService(Supplier<PrestoThriftService> clientSupplier, RetryDriver retry, RetryingPrestoThriftServiceStats retryingPrestoThriftServiceStats)
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: In the caller this parameter is called just `stats`. Let's call it the same here.

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java [hunk 22]
```diff
@@ -85,7 +87,8 @@ void testRetrySuccess()
             throws Exception
     {
         SyncTaskImpl task = new SyncTaskImpl(SYNC_FAILURE, SYNC_FAILURE, SYNC_FAILURE, SYNC_FAILURE, SUCCESS);
-        Integer result = retry.run("test", task);
+        RetryStats retryStats = new RetryStats();
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java]: I think there's value in adding a separate test which checks if stats are populated.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 23]
```diff
@@ -226,4 +247,73 @@ void nextAttempt()
...
+        {
+            finalFailure.update(1);
+        }
+
+        public void addFailureLatency(Duration duration)
+        {
+            failureLatency.add(duration.roundTo(NANOSECONDS));
+        }
+
+        public void addSuccessRetry(int count)
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: I noticed that now you don't record the number of attempts before failure. Why?

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 24]
```diff
@@ -226,4 +247,73 @@ void nextAttempt()
...
+        @Managed
+        @Nested
+        public TimeDistribution getFailureLatency()
+        {
+            return failureLatency;
+        }
+
+        @Managed
+        @Nested
+        public DistributionStat getSuccessRetry()
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: I think it has a meaning of "attempts before success". Let's call it something similar.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java (hunk 23 above)
Comment (hellium01) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Initially, I thought this value will be either 1 or number of retry. But it seems like, the time period between retry is random so this value can be something in between. I will add it.

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java (hunk 22 above)
Comment (hellium01) [reply presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java]: Thanks! added test and found out a bug in counting async retry.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 25]
```diff
@@ -96,11 +101,13 @@ public PrestoThriftService selectedHostClient(List<HostAddress> hosts)
...
         private final RetryDriver retry;
         private PrestoThriftService client;
+        private RetryingPrestoThriftServiceStats retryingPrestoThriftServiceStats;
 
-        public RetryingService(Supplier<PrestoThriftService> clientSupplier, RetryDriver retry)
+        public RetryingService(Supplier<PrestoThriftService> clientSupplier, RetryDriver retry, RetryingPrestoThriftServiceStats stats)
         {
             this.clientSupplier = requireNonNull(clientSupplier, "clientSupplier is null");
             this.retry = retry.onRetry(this::close);
+            this.retryingPrestoThriftServiceStats = requireNonNull(stats, "stats is null");
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: You changed the name of a parameter, but not of the field. Can you also check other places where having a shorter name is preferred?

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 26]
```diff
@@ -143,29 +157,44 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
...
     }
 
-    private <V> ListenableFuture<V> runAsyncInternal(String callableName, Callable<ListenableFuture<V>> callable, RetryStatus retryStatus)
+    private <V> ListenableFuture<V> runAsyncInternal(String callableName, RetryStats stats, Callable<ListenableFuture<V>> callable, RetryStatus retryStatus)
     {
         retryStatus.nextAttempt();
         ListenableFuture<V> resultFuture;
         try {
             resultFuture = callable.call();
+            Futures.addCallback(resultFuture, new FutureCallback()
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: IDE likely gave you warning that you're using generics correctly. Address the warning.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 27]
```diff
@@ -143,29 +157,44 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
...
+                @Override
+                public void onSuccess(@Nullable Object result)
+                {
+                    stats.addSuccess();
+                    stats.addAttemptsBeforeSuccess(retryStatus.getAttempts());
+                    stats.addSuccessLatency(retryStatus.getDuration());
+                }
+
+                @Override
+                public void onFailure(Throwable t)
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: put curly braces on the same line

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 28]
```diff
@@ -226,4 +263,87 @@ void nextAttempt()
             }
         }
     }
+
+    public static final class RetryStats
+    {
+        private final CounterStat success = new CounterStat();
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Let's put fields and getters in certain order. For example: all success, then all failures, or for each group do success then failure.

File: presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java [hunk 29]
```diff
@@ -75,28 +76,39 @@ void testNoRetry()
             throws Exception
     {
         SyncTaskImpl task = new SyncTaskImpl(SUCCESS);
-        Integer result = retry.run("test", task);
+        RetryStats retryStats = new RetryStats();
+        Integer result = retry.run("test", retryStats, task);
 
         assertSyncSuccess(result, task, 1);
+        assertEquals(retryStats.getSuccess().getTotalCount(), 1);
```
Comment (AlekseiS) [inline presto-thrift-connector/src/test/java/com/facebook/presto/connector/thrift/util/TestRetryDriver.java]: This block is exactly the same in all tests. Think of adding a helper method. For example, you can add two: one for success, one for failure.
Then you'll do something like `assertSuccessfulStats(1); // number of attempts` and `assertFailedStats(3)`. Then move 4 lines you have into each of them.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java (hunk 27 above)
Comment (AlekseiS) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: put curly braces on the same line as the method

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 30]
```diff
@@ -226,4 +262,87 @@ void nextAttempt()
...
+        private final CounterStat finalFailure = new CounterStat();
+        private final DistributionStat attemptsBeforeFailure = new DistributionStat();
+        private final TimeDistribution failureLatency = new TimeDistribution(MICROSECONDS);
```
Comment (AlekseiS) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: You use microseconds as a unit, but do round to nanos when you update the metric. How does it work in such case?

**Comment by AlekseiS** [comment]
Oh, and squash commits before merging.

**Comment by hellium01** [comment]
Rebased & Squashed

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java [hunk 31]
```diff
@@ -178,4 +185,56 @@ public void close()
             client = null;
         }
     }
+
+    public static class RetryingPrestoThriftServiceStats
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: This can be named just `Stats` as it is a nested class

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java (hunk 12 above)
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Let’s call this `stats`, as the current name is hard to distinguish from `retryStatus`.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java [hunk 32]
```diff
@@ -174,19 +201,22 @@ public RetryDriver withClassifier(Function<Exception, Exception> classifier)
...
-            return Futures.dereference(retryExecutorService.schedule(() -> runAsyncInternal(callableName, callable, retryStatus), retryStatus.getRetryDelayInMs(), MILLISECONDS));
+            return Futures.dereference(retryExecutorService.schedule(() -> runAsyncInternal(callableName, stats, callable, retryStatus), retryStatus.getRetryDelayInMs(), MILLISECONDS));
         });
     }
 
-    private RuntimeException propagate(Exception e)
+    private RuntimeException propagate(RetryStatus retryStatus, RetryStats stats, Exception e)
     {
         Exception classified = classifier.apply(e);
+        stats.addFinalFailure();
```
Comment (electrum) [inline presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: This is really unrelated to throwing an exception. How about we separate it into a separate method `recordFinalFailure(RetryStatus, RetryStats)`

Comment (electrum) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Also, it's weird to shove this in the middle of where `classified` is computed and used

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java (hunk 30 above)
Comment (electrum) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: Yes, this seems to be broken. You can call `getUnit()` on `TimeDistribution` when doing the `add()`

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftModule.java [hunk 33]
```diff
@@ -54,6 +65,9 @@ public void configure(Binder binder)
         configBinder(binder).bindConfig(ThriftConnectorConfig.class);
         binder.bind(ThriftSessionProperties.class).in(Scopes.SINGLETON);
         binder.bind(ThriftIndexProvider.class).in(Scopes.SINGLETON);
+        binder.bind(RetryingPrestoThriftServiceStats.class).in(Scopes.SINGLETON);
+        newExporter(binder).export(RetryingPrestoThriftServiceStats.class)
+                .as(generatedNameOf(RetryingPrestoThriftServiceStats.class, connectorId));
```
[the remark this answers is not in the document]
Comment (electrum) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftModule.java]: Actually, I'm not sure exporting a nested class name works. You should use a different name. Maybe `PrestoThriftService`?

[the remark this answers is not in the document]
Comment (hellium01) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftModule.java]: I did see the nested class exported, but I think naming it Stats and export on the outer class is cleaner.  I will still change it.

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java (hunk 31 above)
Comment (hellium01) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/clientproviders/RetryingPrestoThriftServiceProvider.java]: Changing it

File: presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java (hunk 30 above)
Comment (hellium01) [reply presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/util/RetryDriver.java]: I think add() takes nano second input and the unit is just for output conversion. In other places of code, we are adding nanosecond into the time distribution.  Testing in test cluster, I see values populated as 70,000 microseconds, so this should be correct?
